framework/src/userobject/ElementSubdomainModifier.C

```diff
@@ -0,0 +1,370 @@
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      _moved_nodes.insert(elem->node_id(i));
+
+    elem->subdomain_id() = subdomain_id;
+    _moved_elems.push_back(elem);
+    if (displaced_elem)
+    {
+      displaced_elem->subdomain_id() = subdomain_id;
+      _moved_displaced_elems.push_back(displaced_elem);
+    }
+  }
+}
+
+void
+ElementSubdomainModifier::finalize()
+{
+  /*
+    Synchronize ghost element subdomain ID
+    Note: this needs to be done before updating boundary info because
+    updating boundary requires the updated element subdomain ids
+  */
+  SyncSubdomainIds sync(_mesh.getMesh());
+  Parallel::sync_dofobject_data_by_id(_mesh.getMesh().comm(),
+                                      _mesh.getMesh().elements_begin(),
+                                      _mesh.getMesh().elements_end(),
+                                      sync);
+  updateBoundaryInfo(_mesh, _moved_elems);
+
+  // Similarly for the displaced mesh
+  if (_displaced_problem)
+  {
+    SyncSubdomainIds sync_displaced(_displaced_problem->mesh().getMesh());
+    Parallel::sync_dofobject_data_by_id(_displaced_problem->mesh().getMesh().comm(),
+                                        _displaced_problem->mesh().getMesh().elements_begin(),
+                                        _displaced_problem->mesh().getMesh().elements_end(),
+                                        sync_displaced);
+    updateBoundaryInfo(_displaced_problem->mesh(), _moved_displaced_elems);
+  }
+
+  // Reinit equation systems
+  _fe_problem.meshChanged();
+
+  // Apply initial condition for the newly moved elements and boundary nodes
+  buildMovedElemsRange();
+  buildMovedBndNodesRange();
+
+  if (_apply_ic)
+  {
+    _fe_problem.projectInitialConditionOnCustomRange(movedElemsRange(), movedBndNodesRange());
+
+    // Set old and older solution on the initialized dofs
+    setOldAndOlderSolutionsForMovedNodes(_fe_problem.getNonlinearSystemBase());
+    setOldAndOlderSolutionsForMovedNodes(_fe_problem.getAuxiliarySystem());
+
+    if (_fe_problem.isTransient())
+      _fe_problem.restoreSolutions();
+  }
+
+  // Initialize stateful material properties for the newly activated elements
+  _fe_problem.initElementStatefulProps(movedElemsRange());
+}
+
+void
+ElementSubdomainModifier::updateBoundaryInfo(MooseMesh & mesh,
+                                             const std::vector<const Elem *> & moved_elems)
+{
+  if (!_moving_boundary_specified)
+    return;
+
+  BoundaryInfo & bnd_info = mesh.getMesh().get_boundary_info();
+
+  // save the removed ghost sides and associated nodes to sync across processors
+  std::unordered_map<processor_id_type, std::vector<std::pair<dof_id_type, unsigned int>>>
+      ghost_sides_to_remove;
+  std::unordered_map<processor_id_type, std::vector<dof_id_type>> ghost_nodes_to_remove;
+
+  // The logic below updates the side set and the node set associated with the moving boundary.
```

Comment:
@fdkong @dschwen @loganharbour I only changed the following logic compared to #17101 .
Here from line 172 to line 221.

As far as I can tell, the parallel issue stems from off-processor point neighbors that only shares one node with the moved element.